src/core/tsi/ssl_transport_security.cc

```diff
@@ -568,21 +572,84 @@ static tsi_result ssl_ctx_use_private_key(SSL_CTX* context, const char* pem_key,
   EVP_PKEY* private_key = nullptr;
   BIO* pem;
   GPR_ASSERT(pem_key_size <= INT_MAX);
-  pem = BIO_new_mem_buf((void*)pem_key, static_cast<int>(pem_key_size));
-  if (pem == nullptr) return TSI_OUT_OF_RESOURCES;
-  do {
-    private_key = PEM_read_bio_PrivateKey(pem, nullptr, nullptr, (void*)"");
-    if (private_key == nullptr) {
-      result = TSI_INVALID_ARGUMENT;
-      break;
-    }
-    if (!SSL_CTX_use_PrivateKey(context, private_key)) {
-      result = TSI_INVALID_ARGUMENT;
-      break;
-    }
-  } while (0);
-  if (private_key != nullptr) EVP_PKEY_free(private_key);
-  BIO_free(pem);
+
+// BoringSSL does not have ENGINE_load_dynamic, ENGINE_by_id
+// support.
+#ifndef OPENSSL_IS_BORINGSSL
+  if (strncmp(pem_key, kSslEnginePrefix, strlen(kSslEnginePrefix)) == 0) {
+    ENGINE* engine;
+    do {
+      char *p, *last;
+      gpr_log(GPR_DEBUG, "ENGINE key specified");
+      p = (char*)pem_key + strlen(kSslEnginePrefix);
```

Comment:
It doesn't seem okay to modify this buffer because `pem_key` is the `const char *` argument.

Reply:
It is just in place parsing. Not changing the string in the end.

Reply:
Even thought all change will be reverted, it doesn't seem right since the string passed here could be in strictly read-only area and this potentially causes segfault.

Reply:
Fixed.